crypto/bio/b_print.c

```diff
@@ -635,10 +636,16 @@ fmtfp(char **sbuffer,
             fvalue = tmpvalue;
     }
     ufvalue = abs_val(fvalue);
-    if (ufvalue > ULONG_MAX) {
-        /* Number too big */
+    /*
+     * Check if the real value is too large to be displayed via an
+     * unsigned long int.
+     *
+     * The obvious comparison ufvalue > ULONG_MAX doesn't work because
+     * ULONG_MAX cannot be converted to a double without accuracy loss.
+     */
+    if (ufvalue >= ldexp(1., sizeof(unsigned long) * 8))
```

Comment:
I hope that function is available everywhere...

Reply:
It's present in C89 and I think it was in K&R.

What I'm not certain about is compilers optimising away the call and avoiding libm being a dependency.

Reply:
did you also consider `if (ufvalue >= 18446744073709551616.0) {` ?

Reply:
Hmm, we don't know for sure what ULONG_MAX is, but would this work?
`if (ifValue >= (double)(ULONG_MAX - 65535) + 65536.0) {`

Idea: ULONG_MAX-65535 cancels 16 bits, so that should be
representable as a 53 bit float and adding 65536.0 is also an exact result.

Reply:
C89?  You sure? "    The ldexp() functions conform to ISO/IEC 9899:2011."

Adding -lm as a dependency is a deal-breaker, IMO.

Reply:
Yes, I did some checking and ldexp is in v7 unix but in -lm.

Maybe change the docs/code to say "doesn't work within 1K of MAX values"

Reply:
I would also like to avoid a new linker dependency whatever it costs.
Especially when this should be back-ported to the 1.1.1 LTS branch.
You must at least subtract 2K-1 from ULONG_MAX, to get the next lower
double number.  I think IEEE double has 53 bit mantissa, while u_long has 64 bit, so
64 bit - 53 bit = 11, so 2K = 2^11 would make sense.
FYI my last comment was not meant as a joke ;-)
I think that should really do the trick.

Reply:
> C89? You sure? " The ldexp() functions conform to ISO/IEC 

Different manuals say different things.
The OpenBSD manual says C89: https://man.openbsd.org/ldexp.3
The Linux manual says C99 and POSIX-1: http://man7.org/linux/man-pages/man3/ldexp.3.html

Reply:
It was in V7, according to https://s3.amazonaws.com/plan9-bell-labs/7thEdMan/vol1/man3.bun.txt

Reply:
> did you also consider `if (ufvalue >= 18446744073709551616.0) {`

Yes.


I do remember the `ldexp()` function being around very early.  I don't have a copy of K&R anymore so couldn't check.  The _float_ and _long double_ versions are far more recent.